src/sgame/sg_bot.cpp

```diff
@@ -263,6 +262,125 @@ bool G_BotSetDefaults( int clientNum, team_t team, int skill, Str::StringRef beh
 	return true;
 }
 
+struct botSkillTreeElement_t {
+	Str::StringRef name;
+	enum bot_skill enable;
+	int cost;
+	// function that determines if this skill is available
+	std::function<bool(const gentity_t *self, skillSet_t existing_skills)> predicate;
+	// skills that are unlocked once you unlock this one
+	std::vector<botSkillTreeElement_t> unlocked_skills;
+};
+
+static const std::vector<botSkillTreeElement_t> initial_unlockable_skills = {
+};
+
+// Note: this function modifies possible_choices to remove the one we just
+//       chose, and add the new unlocked skills to that list.
+static Util::optional<botSkillTreeElement_t> ChooseOneSkill(const gentity_t *bot, skillSet_t skillSet, std::vector<botSkillTreeElement_t> &possible_choices, std::mt19937_64& rng)
+{
+	int total_skill_points = 0;
+	for (const auto &skill : possible_choices)
+	{
+		if (skill.predicate(bot, skillSet))
+		{
+			total_skill_points += skill.cost;
+		}
+	}
+
+	if (total_skill_points <= 0)
+		return Util::nullopt;
+
+	std::uniform_int_distribution<int> dist(0, total_skill_points - 1);
+	int random = dist(rng);
+
+	// find which skill this corresponds to (higher point skills are more
+	// likely to be found)
+	int cursor = 0;
+	for (std::vector<botSkillTreeElement_t>::iterator skill = possible_choices.begin(); skill != possible_choices.end(); skill++)
+	{
+		if (!skill->predicate(bot, skillSet))
+		{
+			continue;
+		}
+
+		cursor += skill->cost;
+		if (random < cursor)
+		{
+			botSkillTreeElement_t result = *skill;
+
+			// update the list of future choices
+			possible_choices.erase(skill);
+			for (const auto &child : result.unlocked_skills)
+			{
+				possible_choices.push_back(std::move(child));
+			}
+			result.unlocked_skills.clear();
+
+			return result;
+		}
+	}
+	return Util::nullopt;
+}
+
+// This selects a set of skills for a bot.
+//
+// The algorithm is stable and deterministic.
+// Deterministic:
+//     a bot's skill set is determined only by the bot's name and the bot's
+//     skill level. A different bot will have a different skillset, but this
+//     means that a bot with the same name will have the same skillset from
+//     a game to another.
+// Stable:
+//     when a bot changes skill level, the first skills unlocked are common
+//     between both skill levels. This means that a bot whose skill is raised
+//     by BotSetSkillLevel will keep the skills it already had, and get new
+//     ones.
+//     This essentially means that a bot will keep its "personality" after
+//     a skill change, although keeping its base skills or unlocking new ones.
+std::pair<std::string, skillSet_t> BotDetermineSkills(gentity_t *bot, int skill)
+{
+	std::vector<botSkillTreeElement_t> possible_choices = initial_unlockable_skills;
+	int skill_points = skill * 4; // FIXME
+
+	// rng preparation
+	std::string name = bot->client->pers.netname;
+	std::seed_seq seed(name.begin(), name.end());
+	std::mt19937_64 rng(seed);
```

Comment:
Use a lower-quality random number generator. This has an unnecessarily huge state.

Reply:
Heh, I didn't do it in the end, but I wanted to write "performance is a non-goal" on the header of this function.

I initially used a 32bits version of the mersenne twister, but had skill which were too similar from a run to another. I don't know if it was just bad luck and if that would still be a problem given that there are more possible skills now.

Reply:
I still think it would be nice to use a generator that doesn't have a 2.5 kB size.

Reply:
Maaaaybe ? But also I tested this one and the generator is destroyed once bot have finished their initialisation step so it really doesn't matter?

Reply:
Well that's kind of the point; you're initializing this expensive object and then immediately throwing it away.

Why is this so hard? AFAIU it's just find-and-replace of mt19937 with the name of some other generator. If you use a default configuration and don't try to make up your own parameters there shouldn't be any danger.

Reply:
Well the reason I'm not so fond of that change is that I had an impression that the previous RNG did give a bad result. The randomness looked not-so-random. So I switched to a fancier generator. Maybe that was just a feeling due to sample size, I don't know.

About how hard is it to change, well I have to edit things twice because the code was moved to another file and ensure it still builds mid-branch to avoid being annoying on bisect, plus some testing. It's a moderately important cost that I don't feel is needed.

If we want to make bot loading faster anyway, we have to look at the navmesh loading — that's what actually takes time.

Reply:
You can add a new commit at the end rather than rebasing.